test/brain/what_if_index_test.cpp

```diff
@@ -0,0 +1,488 @@
+
+  binder->BindNameToNode(sql_statement.get());
+
+  // Get the optimized plan tree without the indexes (sequential scan)
+  auto result = brain::WhatIfIndex::GetCostAndBestPlanTree(
+      sql_statement, config, DEFAULT_DB_NAME, txn);
+  auto cost_without_index = result->cost;
+  LOG_DEBUG("Cost of the query without indexes {}: %lf", cost_without_index);
+  LOG_DEBUG("%s", result->plan->GetInfo().c_str());
+
+  // Insert hypothetical catalog objects
+  config.AddIndexObject(
+      testing_util.CreateHypotheticalIndex(schema.table_name, {"a", "c"}));
+
+  result = brain::WhatIfIndex::GetCostAndBestPlanTree(sql_statement, config,
+                                                      DEFAULT_DB_NAME, txn);
+  auto cost_with_index_1 = result->cost;
+  LOG_DEBUG("Cost of the query with index {'a', 'c'}: %lf", cost_with_index_1);
+  EXPECT_EQ(result->plan->GetPlanNodeType(), PlanNodeType::SEQSCAN);
+  EXPECT_DOUBLE_EQ(cost_without_index, cost_with_index_1);
+  LOG_DEBUG("%s", result->plan->GetInfo().c_str());
+
+  config.Clear();
+  config.AddIndexObject(
+      testing_util.CreateHypotheticalIndex(schema.table_name, {"a", "b"}));
+  result = brain::WhatIfIndex::GetCostAndBestPlanTree(sql_statement, config,
+                                                      DEFAULT_DB_NAME, txn);
+  auto cost_with_index_2 = result->cost;
+  LOG_DEBUG("Cost of the query with index {'a', 'b'}: %lf", cost_with_index_2);
+  EXPECT_EQ(result->plan->GetPlanNodeType(), PlanNodeType::SEQSCAN);
+  EXPECT_DOUBLE_EQ(cost_without_index, cost_with_index_2);
+  LOG_DEBUG("%s", result->plan->GetInfo().c_str());
+
+  config.Clear();
+  config.AddIndexObject(
+      testing_util.CreateHypotheticalIndex(schema.table_name, {"b", "c"}));
+  result = brain::WhatIfIndex::GetCostAndBestPlanTree(sql_statement, config,
+                                                      DEFAULT_DB_NAME, txn);
+  auto cost_with_index_3 = result->cost;
+  LOG_DEBUG("Cost of the query with index {'b', 'c'}: %lf", cost_with_index_3);
+  EXPECT_EQ(result->plan->GetPlanNodeType(), PlanNodeType::INDEXSCAN);
+  EXPECT_GT(cost_without_index, cost_with_index_3);
+  LOG_DEBUG("%s", result->plan->GetInfo().c_str());
+
+  config.Clear();
+  config.AddIndexObject(
+      testing_util.CreateHypotheticalIndex(schema.table_name, {"b"}));
+  result = brain::WhatIfIndex::GetCostAndBestPlanTree(sql_statement, config,
+                                                      DEFAULT_DB_NAME, txn);
+  auto cost_with_index_4 = result->cost;
+  EXPECT_LE(cost_with_index_3, cost_with_index_4);
+
+  // The cost of using one index {1} should be greater than the cost
+  // of using both the indexes {1, 2} for the query.
+  EXPECT_EQ(result->plan->GetPlanNodeType(), PlanNodeType::INDEXSCAN);
+  LOG_DEBUG("Cost of the query with index {'b'}: %lf", cost_with_index_4);
+  LOG_DEBUG("%s", result->plan->GetInfo().c_str());
+
+  txn_manager.CommitTransaction(txn);
+}
+
+TEST_F(WhatIfIndexTests, MultiColumnTest2) {
```

Comment:
What is the exact difference between the different MultiColumnTests - it seems like its testing different workloads with different table setups and configurations. Is this the case or is it something specific - Might be useful to have brief comments?

Reply:
One of them is a very basic test. The second one is a more complex test with just one query in the workload. The third one uses workload with multiple queries.